Attempt to deflake DBLogicalBlockSizeCacheTest.CreateColumnFamilies

hx235 opened the pull request:
**Context:**
`DBLogicalBlockSizeCacheTest.CreateColumnFamilies` is flaky on a rare occurrence of assertion failure below
```
db/db_logical_block_size_cache_test.cc:210
Expected equality of these values:
  1
  cache_->GetRefCount(cf_path_0_)
    Which is: 2
```

Root-cause: `ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[0]));` in the test may not successfully decrease the ref count of `cf_path_0_` since the decreasing only happens in the clean-up of `ColumnFamilyData` when `ColumnFamilyData` has no referencing to it, which may not be true when `db->DestroyColumnFamilyHandle(cfs[0])` is called since background work such as `DumpStats()` can hold reference to that `ColumnFamilyData` (suggested and repro-d by @ajkr ). Similar case `ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[1]));`.

See following for a deterministic repro:
```
diff --git a/db/db_impl/db_impl.cc b/db/db_impl/db_impl.cc
index 196b428a3..4e7a834c4 100644
--- a/db/db_impl/db_impl.cc
+++ b/db/db_impl/db_impl.cc
@@ -956,10 +956,16 @@ void DBImpl::DumpStats() {
         // near-atomically.
         // Get a ref before unlocking
         cfd->Ref();
+        if (cfd->GetName() == "cf1" || cfd->GetName() == "cf2") {
+          TEST_SYNC_POINT("DBImpl::DumpStats:PostCFDRef");
+        }
         {
           InstrumentedMutexUnlock u(&mutex_);
           cfd->internal_stats()->CollectCacheEntryStats(/*foreground=*/false);
         }
+        if (cfd->GetName() == "cf1" || cfd->GetName() == "cf2") {
+          TEST_SYNC_POINT("DBImpl::DumpStats::PreCFDUnrefAndTryDelete");
+        }
         cfd->UnrefAndTryDelete();
       }
     }
diff --git a/db/db_logical_block_size_cache_test.cc b/db/db_logical_block_size_cache_test.cc
index 1057871c9..c3872c036 100644
--- a/db/db_logical_block_size_cache_test.cc
+++ b/db/db_logical_block_size_cache_test.cc
@@ -9,6 +9,7 @@
 #include "env/io_posix.h"
 #include "rocksdb/db.h"
 #include "rocksdb/env.h"
+#include "test_util/sync_point.h"
 
 namespace ROCKSDB_NAMESPACE {
 class EnvWithCustomLogicalBlockSizeCache : public EnvWrapper {
@@ -183,6 +184,15 @@ TEST_F(DBLogicalBlockSizeCacheTest, CreateColumnFamilies) {
   ASSERT_EQ(1, cache_->GetRefCount(dbname_));
 
   std::vector<ColumnFamilyHandle*> cfs;
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->LoadDependency(
+      {{"DBLogicalBlockSizeCacheTest::CreateColumnFamilies::PostSetupTwoCFH",
+        "DBImpl::DumpStats:StartRunning"},
+       {"DBImpl::DumpStats:PostCFDRef",
+        "DBLogicalBlockSizeCacheTest::CreateColumnFamilies::PreDeleteTwoCFH"},
+       {"DBLogicalBlockSizeCacheTest::CreateColumnFamilies::"
+        "PostFinishCheckingRef",
+        "DBImpl::DumpStats::PreCFDUnrefAndTryDelete"}});
   ASSERT_OK(db->CreateColumnFamilies(cf_options, {"cf1", "cf2"}, &cfs));
   ASSERT_EQ(2, cache_->Size());
   ASSERT_TRUE(cache_->Contains(dbname_));
@@ -190,7 +200,7 @@ TEST_F(DBLogicalBlockSizeCacheTest, CreateColumnFamilies) {
   ASSERT_TRUE(cache_->Contains(cf_path_0_));
   ASSERT_EQ(2, cache_->GetRefCount(cf_path_0_));
   }
 
    // Delete one handle will not drop cache because another handle is still
   // referencing cf_path_0_.
+  TEST_SYNC_POINT(
+      "DBLogicalBlockSizeCacheTest::CreateColumnFamilies::PostSetupTwoCFH");
+  TEST_SYNC_POINT(
+      "DBLogicalBlockSizeCacheTest::CreateColumnFamilies::PreDeleteTwoCFH");
   ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[0]));
   ASSERT_EQ(2, cache_->Size());
   ASSERT_TRUE(cache_->Contains(dbname_));
@@ -209,16 +221,20 @@ TEST_F(DBLogicalBlockSizeCacheTest, CreateColumnFamilies) {
   ASSERT_TRUE(cache_->Contains(cf_path_0_));
    // Will fail
   ASSERT_EQ(1, cache_->GetRefCount(cf_path_0_));
 
   // Delete the last handle will drop cache.
   ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[1]));
   ASSERT_EQ(1, cache_->Size()); 
   ASSERT_TRUE(cache_->Contains(dbname_));
   // Will fail
   ASSERT_EQ(1, cache_->GetRefCount(dbname_));
 
+  TEST_SYNC_POINT(
+      "DBLogicalBlockSizeCacheTest::CreateColumnFamilies::"
+      "PostFinishCheckingRef");
   delete db;
   ASSERT_EQ(0, cache_->Size());
   ASSERT_OK(DestroyDB(dbname_, options,
       {{"cf1", cf_options}, {"cf2", cf_options}}));
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
 }
```

**Summary**
- Removed the flaky assertion
- Clarified the comments for the test


**Test**
- Run repro above with no assertion failure
- CI
- Monitor for future flakiness

ajkr reviewed (changes requested):
Sorry this doesn't leave us with a more understandable/maintainable test compared to before. Why do we need to disable DumpStats() and loosen some refcount checks? The description sounds like we gave up on repro - did the hypothesis not lead to any progress?

hx235 commented:
> Sorry this doesn't leave us with a more understandable/maintainable test compared to before. Why do we need to disable DumpStats() and loosen some refcount checks? The description sounds like we gave up on repro - did the hypothesis not lead to any progress?

Replied in the task

hx235 commented:
Discussed offline that we are going to either loosen the assertion without disabling the dump stat or verifying it's actually the dump stat reference and remove the dump stat. Not both the same time in this PR for clarity in testing.

hx235 commented:
Update:
- Reverted changes related to DumpStat()
- Clarified comment of indeterministic CF ref count
- Replied internally on keeping the exact refcount assertion on db ref

ajkr commented on db/db_logical_block_size_cache_test.cc:
```diff
@@ -188,32 +188,45 @@ TEST_F(DBLogicalBlockSizeCacheTest, CreateColumnFamilies) {
   ASSERT_EQ(1, cache_->GetRefCount(dbname_));
   ASSERT_TRUE(cache_->Contains(cf_path_0_));
-  ASSERT_EQ(2, cache_->GetRefCount(cf_path_0_));
+  // We can't assert exact ref count of column family
+  // as there might be extra reference to the column family
+  // outside of CF creation before we delete the CF, which causes the exact
+  // assertion to be flaky.
+  ASSERT_GE(2, cache_->GetRefCount(cf_path_0_));
```
As mentioned offline I think these `ASSERT_GE(2, ...`s can be reverted. Extra references to ColumnFamilyData (e.g., from DumpStats()) do not increase the number of calls to `RegisterDbPaths()`

hx235 replied:
For clarification, I was trying to assert there are at least two ref, one for each cf handle
Are you suggesting to revert this particular `ASSERT_GE(2,..)` to exact assertion, OR all the `ASSERT_GE(2, ...) to exact assertion OR completely removing `ASSERT_GE(2, ...)`?

hx235 replied:
For (a), are we sure there won't be any extra reference to the cf other than the 2 references from the handles?
- I am not quite sure about this especially after we see some in the following places without reproducing it.

For (b), we will probably need to keep the `ASSERT_GE(2, ...)` at least for the place where we've seen the violation.

F(c), I think at least we need to assert the minimum of the column ref (even though that might be a false positive due to "extra reference" being a lot. 

For "(d)", we just do `ASSERT_GE(0, ...)` for all places other than deletion. But that will prevent us from testing the case for "Delete one handle will not drop CF's entry from cache because another handle is still referencing cf_path_0_."

ajkr replied:
> For clarification, I was trying to assert there are at least two ref, one for each cf handle

The calls to `RegisterDbPaths()` and `UnregisterDbPaths()` are in the `ColumnFamilyData` constructor/destructor, not `ColumnFamilyHandle`.

> Just to make sure I understand, are you suggesting to (a) revert this particular ASSERT_GE(2,..) to exact assertion, OR (b) all the ASSERT_GE(2, ...) to exact assertion OR (c) completely removing ASSERT_GE(2, ...)?

(b)

ajkr replied:
> > For clarification, I was trying to assert there are at least two ref, one for each cf handle
> The calls to RegisterDbPaths() and UnregisterDbPaths() are in the ColumnFamilyData constructor/destructor, not ColumnFamilyHandle.

To clarify further, I realize you did not say that ColumnFamilyHandle references/unreferences the cache entry. I just have no idea how there could ever be more than two references here or the other place that says `ASSERT_GE(2, ...`.

ajkr replied:
> For (b), we will probably need to keep the ASSERT_GE(2, ...) at least for the place where we've seen the violation.

You do not need to keep any ASSERT_GE(2, ...) because the violation is avoided by changing the `ASSERT_EQ(1, ...)` to `ASSERT_GE(1, ...)`. That place the change is justified because the dropped `ColumnFamilyData`'s lifetime could have been extended by `DumpStats()`, e.g.

hx235 replied:
> To clarify further, I realize you did not say that ColumnFamilyHandle references/unreferences the cache entry. I just have no idea how there could ever be more than two references here or the other place that says ASSERT_GE(2, ....

Hmmm let me try to clarify myself a bit more. The original (flaky) test did `ASSERT_EQ(2, cache_->GetRefCount(cf_path_0_));` when the CF is created with two handles "cf1", "cf2" and does  `ASSERT_EQ(1, cache_->GetRefCount(cf_path_0_));` after one handle is dropped. I thought two handles corresponding to two refcount of `cf_path_0` is something I have to accept. 

Then since we see some flakiness in  `ASSERT_EQ(1, cache_->GetRefCount(cf_path_0_));` revealing there might some extra reference to the `cf_path_0` that we are not sure what it is without a clear repro. Maybe it's DumpStat() or maybe it's something else or maybe it's something more (e.g, like adding 2 extra reference to the existing 2). And I don't really know what that "something else/more" is 
and when that "something else/more" could happen. It could happen from the beginning we create the column family.  

So that's why I might have been conservative in asserting ASSERT_GE(2, ..).

Would you prefer us assuming the extra "unconfirmed" reference only increase the ref count by 1 and only happen at the place we see the violation?

ajkr replied:
I'll try to repro and submit my idea of the fix upon returning since I disagree with the approach (loosening assertions because there might be something we haven't thought of).

hx235 replied:
> (loosening assertions because there might be something we haven't thought of).

Well that's just my thought I'd like to share so that I can clearly get to where we disagree. I also wouldn't mind reverting those ASSERT_GE() to be honest. All I said above was just trying to make my point of view clear and again get to where we disagree, not necessarily asking for agreement :) Will update the PR. 

I also welcome any efforts into reproducing it as this deserves more time than I have spent on it so far.

ajkr replied:
> Edit: I guess the underlying understanding I have from the test being flaky is that there can be extra refcount to cf_path_0 other than the ones from cf handles in the creation. I don't know what those ref count are/when they start to appear. Me not knowing it might be a result of not investigating the code enough (e.g, pinning down exactly where in the code places can ref cf_path_0)

The only code path I can find that references cf_path_0 is `ColumnFamilyData` constructor -> `RegisterDbPaths()` -> `RefAndCacheLogicalBlockSize()`. I believe only the two `ColumnFamilyData`s (the one for "cf1" and the one for "cf2") register cf_path_0.

On the other hand, `ColumnFamilyData` could be referenced in more places, like for background work. Extra references to `ColumnFamilyData` extend its lifetime beyond when it is dropped and the user's CF handle is released. The effect is just it takes longer than the test expects to release a reference to cf_path_0 via `ColumnFamilyData` destructor -> `UnregisterDbPaths()` -> `UnrefAndTryRemoveCachedLogicalBlockSize()`.

ajkr replied:
> I also welcome any efforts into reproducing it as this deserves more time than I have spent on it so far.

I think we should still reconsider who should be the PR author. We shouldn't be doing reviewer-driven development. It is quite inefficient and took many rounds of back and forth for me to even realize that the `ASSERT_GE(2, ...` changes were based on not investigating it yet.

hx235 replied:
Thanks for the repro! For other comments you left, I will reply through internal chat but no pressure in checking in before you return :)

hx235 commented:
Update: 
- Removed extra assertion loosening and fixed only the places based on the evidence of where the flakiness happen
- Ready for review after the signal is clear

tangrames commented:
      谢谢，您的邮件我已经收到，看到邮件后，我会第一时间回复您！
      祝一切顺利！

hx235 commented:
Update:
- Clarified the testing in comments and **removed** (instead of loosening) the flaky assertion for readability.

ajkr reviewed (approved):
LGTM, thanks!
